### ISSUE TITLE
minor Laplace discrepancies with toenail data

### ISSUE THREAD

**Issue by bbolker** [lme4/lme4#289]
This is discussed in somewhat more detail at http://rpubs.com/bbolker/toenail_laplace , but: there's a difference among R packages that implement Laplace approximation on the (in)famous toenail data.  The discrepancy is large enough that I think there's a real issue, although it may not be practically important.

```
##                          RE_sd   (Intercept)  treatment      visit                logLik elapsed
## glmer           4.687377   -1.051835 -0.6968895 -0.9115262 -626.1709     1.076
## glmmML      4.708440   -1.070209 -0.7005596 -0.9126392 -626.1627      0.101
## glmmADMB 4.723452   -1.070187 -0.7005568 -0.9126386 -626.1630 112.825
## SAS              4.708440   -1.070200 -0.7005000 -0.9126000 -626.1650     1.180
```

Basic code:
```
library("foreign") ## for read.dta
library("lme4")
library("glmmML")
library("glmmADMB")
toenail <- read.dta("http://www.stata-press.com/data/r12/toenail.dta")
## summary(toenail)
toenail <- transform(toenail,
                     patient=factor(patient),
                     ## glmmML, glmmADMB prefer numeric 0/1 responses
                     n.outcome=as.numeric(outcome)-1)
t_glmer <- system.time(
    m_glmer <- glmer(outcome~treatment+visit+(1|patient),toenail,
            family=binomial))
t_glmmML <- system.time(
    m_glmmML <- glmmML(n.outcome~treatment+visit,
                       cluster=toenail$patient,toenail,
                       family="binomial"))
t_glmmADMB <- system.time(
    m_glmmADMB <- glmmadmb(n.outcome~treatment+visit+(1|patient),toenail,
                           family="binomial"))
```

**Comment by stevencarlislewalker** [comment]
The fact that this is a canonical link example makes this even more surprising.  `glmer` with _non-canonical_ links uses an approximation to the Laplace approximation, but for _canonical_ links the actual Laplace approximation is used.

**Comment by stevencarlislewalker** [comment]
Perhaps it is just a minor issue with how the default nonlinear optimizer handles this particular example.  If you skip the `nAGQ = 0` initial step, then the log-likelihood starts to look more like the other packages:
```{r}
> logLik(glmer(outcome~treatment+visit+(1|patient),toenail,
+              family=binomial, control = glmerControl(nAGQ0initStep = FALSE)))
'log Lik.' -626.1632 (df=4)
```

**Comment by stevencarlislewalker** [comment]
After reading the rpubs link above I'm even more confused, because apparently the optimizer was right.

**Comment by bbolker** [comment]
Yeah, I realized in thinking about that it was indeed a canonical link.

**Comment by stevencarlislewalker** [comment]
But then why does skipping `nAGQ = 0` work I wonder?

**Comment by pawlenartowicz** [comment]
The discrepancy looks like a PIRLS convergence tolerance effect. Raising `tolPwrss` from its default of `1e-7` moves `glmer` onto the values the other packages report.

```r
library(foreign); library(lme4)
toenail <- transform(read.dta("http://www.stata-press.com/data/r12/toenail.dta"),
                     patient = factor(patient))
f <- outcome ~ treatment + visit + (1 | patient)

m0 <- glmer(f, toenail, family = binomial)
m1 <- glmer(f, toenail, family = binomial, control = glmerControl(tolPwrss = 1e-13))
```

| | RE sd | (Intercept) | treatment | visit | logLik |
|-|-|-|-|-|-|
| `glmer` default (`1e-7`) | 4.687377 | −1.051835 | −0.696889 | −0.911526 | −626.170912 |
| `glmer` `tolPwrss=1e-9` | 4.708408 | −1.070181 | −0.700564 | −0.912635 | −626.162768 |
| `glmer` `tolPwrss=1e-13` | 4.708415 | −1.070192 | −0.700558 | −0.912635 | −626.162768 |
| fastglmm (independent) | 4.708414 | −1.070185 | −0.700557 | −0.912636 | −626.162768 |
| glmmML (this thread) | 4.708440 | −1.070209 | −0.700560 | −0.912639 | −626.1627 |
| glmmADMB (this thread) | 4.723452 | −1.070187 | −0.700557 | −0.912639 | −626.1630 |
| SAS (this thread) | 4.708440 | −1.070200 | −0.700500 | −0.912600 | −626.1650 |

The default fit's own log-likelihood is 0.0082 below what it reaches at a tighter tolerance, so it is stopping short of its optimum rather than finding a different one. Neither fit emits a convergence warning. Sweeping the tolerance puts the transition between `1e-8` and `1e-9`: `1e-8` still gives the default's values to 8 figures. Skipping the initial step appears to start PIRLS somewhere that gets closer before the loose tolerance stops it, rather than addressing the cause.

Versions used: R 4.5.3, lme4 1.1.38, fastglmm 0.2.0

**Comment by bbolker** [comment]
Hmm. This is a good point, but I wonder about the global cost of tightening the `tolPwrss` default ...
